### ISSUE TITLE
Improve build times by parallizing manifest update and jar signing tasks

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3440]
The bulk of the build times are from these two tasks. By doing them in parallel we can lower the current build time by more than half.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3544
Reporter: wayneh
Created: 2014-12-05T15:43:18.000-0800

**Comment by rbeckman-nextgen** [comment]
While we're at it, let's stop including ant related jars in the build (ie. ant-header.jar)

Imported Comment. Original Details:
Author: wayneh
Created: 2014-12-05T15:49:01.000-0800

**Comment by rbeckman-nextgen** [comment]
Seems to be working well so far. The default thread count for both is set to 4. Users can set their own number of thread by setting the "manifest_thread_count" or "signjar_thread_count" properties when running the build file. The build now typically takes less than half the time it used to.

Imported Comment. Original Details:
Author: wayneh
Created: 2014-12-15T08:29:33.000-0800

**Comment by rbeckman-nextgen** [comment]
I found that the thing that slowed down builds the most for me (on Linux) was the kernel entropy pool drying up.

jarsigner uses /dev/random on Linux, which just doesn't have a lot of volume on modern systems with lots of RAM, because it sources entropy (by default) from disk, mouse, and keyboard events.

I've had build times in excess of 30 minutes on my modern quad core workstation, while it was in use. At first I was blaming the timestamp authority because the processes were just sitting there not consuming very much CPU. I hear that virtual machines can really run out of entropy very rapidly because they don't have real disks, so Linux build servers might be very slow building the Mirth distribution.

I made a similar parallel execution patch as linked here but was still seeing very long build times. What fixed it was adding something to inject additional entropy into /dev/random ; I chose to use "randomsound", which samples your microphone for white noise and uses that as entropy, but other approaches like copying a disk to /dev/null to generate disk controller events would also work. If you do this stuff a lot or run a build server and want "enterprise" randomness you might consider getting a hardware RNG.

I have no idea how this affects building Mirth on Windows or OS/X though..

Times from a 2.4GHz i7 (quad core) with 8GB of RAM running Ubuntu 14.04, under normal desktop usage during the build

<parallel>
r7173 : > 31 minutes
r7173 w. randomsound :  4m17s

<for parallel="true">
r7515 : 27m36s
r7515 w. randomsound : 1m53s

Hope that's helpful to other Linux people running into this (also ran into this generating high bit-value GPG keys).

Imported Comment. Original Details:
Author: awilkins
Created: 2015-01-06T06:33:49.000-0800
